lib/context_adapter_communication.js

```diff
@@ -0,0 +1,677 @@
+      };
+      request.payload.contextElements[0].attributes.forEach(function(attribute) {
+        switch (attribute.name) {
+          case caConfig.BUTTON_ENTITY.CA_INTERACTION_TYPE_ATTR_NAME:
+            operationDescriptor.interactionType = attribute.value;
+            break;
+          case caConfig.BUTTON_ENTITY.CA_SERVICE_ID_ATTR_NAME:
+            operationDescriptor.serviceId = attribute.value;
+            break;
+          case caConfig.BUTTON_ENTITY.CA_OPERATION_ACTION_ATTR_NAME:
+            operationDescriptor.action = attribute.value;
+            break;
+          case caConfig.BUTTON_ENTITY.CA_OPERATION_EXTRA_ATTR_NAME:
+            operationDescriptor.extra = attribute.value;
+            break;
+        }
+      });
+      request.contextAdapter.operationDescriptor = operationDescriptor;
+      return process.nextTick(callback.bind(null, null, operationDescriptor));
+    }
+  });
+}
+
+/**
+ * Checks that an queryContext response returned by the Context Adapter to get
+ *  service information is well-formed
+ * @param {object} context The context for logging
+ * @param {object} response The response received from the Context Adapter
+ * @param {string} serviceId The identifier of the service whose information was retrieved
+ * @param {function} callback Callback to be called with the result of the check
+ * @return {*} True if the queryContextResponse is valid. False otherwise.
+ */
+function checkQueryContextResponse(context, response, serviceId, callback) {
+  var queryContextJSONSchema = {
+    properties: {
+      body: {
+        id: 'body',
+        type: 'object',
+        required: true,
+        properties: {
+          contextResponses: {
+            id: 'contextResponses',
+            type: 'array',
+            required: true,
+            minItems: 0,
+            maxItems: 2,
+            items: {
+              id: '0',
+              type: 'object',
+              required: true,
+              properties: {
+                contextElement: {
+                  id: 'contextElement',
+                  type: 'object',
+                  required: true,
+                  properties: {
+                    id: {
+                      id: 'id',
+                      type: 'string',
+                      required: true
+                    },
+                    type: {
+                      id: 'type',
+                      type: 'string',
+                      required: true
+                    },
+                    isPattern: {
+                      id: 'isPattern',
+                      type: 'string',
+                      required: true,
+                      pattern: 'false'
+                    },
+                    attributes: {
+                      id: 'attributes',
+                      type: 'array',
+                      required: true,
+                      minItems: 6,
+                      maxItems: 8,
+                      conform: function(attributes) {
```

Comment:
Extract this function to an external function, it seems weird to define a synchronous function inside a literal object declaration.